sentry/src/main/java/io/sentry/connection/UdpConnection.java

```diff
@@ -0,0 +1,187 @@
+    public void send(Event event) throws ConnectionException {
+        try {
+
+            this.init();
+
+            ByteArrayOutputStream destination = new ByteArrayOutputStream();
+            marshaller.marshall(event, destination);
+            byte[] eventBuf = destination.toByteArray();
+
+            List<String> chunks = new ArrayList<>();
+            chunkify(new String(eventBuf), chunks);
+            int totalChunks = chunks.size();
+            for (int numChunk = 0; numChunk < totalChunks; numChunk++) {
+                String chunk = chunks.get(numChunk);
+                String eventId = event.getId().toString();
+                byte[] sendArray = this.getChunkBytes(chunk, eventId, numChunk + 1, totalChunks);
+                DatagramPacket packet = new DatagramPacket(sendArray, sendArray.length, address, this.port);
+                this.sendPacket(packet);
+            }
+
+            for (EventSendCallback eventSendCallback : eventSendCallbacks) {
+                eventSendCallback.onSuccess(event);
+            }
+
+        } catch (IOException ioex) {
+            for (EventSendCallback eventSendCallback : eventSendCallbacks) {
+                eventSendCallback.onFailure(event, ioex);
+            }
+            throw new ConnectionException("Error writing to udp", ioex);
+        }
+    }
+
+    /**
+     * Package private method to allow mocking of socket send.
+     * @param packet DataGram Packet to send.
+     * @throws IOException can be thrown if socket send fails.
+     */
+    void sendPacket(DatagramPacket packet) throws IOException {
+        this.socket.send(packet);
+    }
+
+    /**
+     * Package private method to allow unit tests to set maxChunkSize.
+     * @param maxChunkSize Max allowable size of chunk.
+     */
+    void setMaxChunkSize(int maxChunkSize) {
+        this.maxChunkSize = maxChunkSize;
+    }
+
+    private void chunkify(String payload, List<String> chunks) {
+        int payloadLength = payload.length();
+        if (payloadLength <= this.maxChunkSize) {
+            chunks.add(payload);
+        } else {
+            int split = payloadLength / 2;
+            String firstHalf = payload.substring(0, split);
+            String secondHalf = payload.substring(split);
+            chunkify(firstHalf, chunks);
+            chunkify(secondHalf, chunks);
+        }
+    }
+
+    private byte[] getChunkBytes(String eventBuff, String eventId, int numChunk, int total) throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        JsonGenerator generator = this.jsonFactory.createGenerator(baos);
+        generator.writeStartObject();
+        generator.writeStringField("id", eventId);
+        generator.writeNumberField("numChunk", numChunk);
+        generator.writeNumberField("total", total);
+        generator.writeStringField("chunk", eventBuff);
+        generator.writeEndObject();
+        generator.flush();
+        return baos.toByteArray();
```

Comment:
Out of curiosity is this an existing "protocol" you use or is this a new pattern some downstream consumer has to understand? I assume it's a new one-off thing?

Reply:
On second thought (unless I'm missing something) if this has no meaning to other users or to the Sentry server I don't think it belongs in the SDK codebase. If this is an internal protocol you're going to support it belongs in a subclass you manage, I think.

Reply:
@bretthoerner It's true that the receiver will need to implement chunk re-assembly. This approach is pretty standard in general. My goal here is for my users to be able to reference the standard SDK from Sentry as opposed to a "custom" fork. If you still think this does not belong here, I just realized that rsyslog also supports tcp transport, and I can remodel this as a TcpConnection, where no chunking is necessary. How does that sound?

Reply:
Is there any reason the UDP/custom implementation cant live in its own class, in another repo/codebase? You can still use the Java SDK apis and bind the transport, right?

Reply:
Yes it can, but then it requires the users to take two dependencies as opposed to using the standard sentry sdk. I am OK to move the UDP implementation out, but then are you guys open to the TCP route?

Reply:
Would the Sentry SDK not just be a transitive dependency of your code, so they'd just require your dep and be done?

I think this probably only belongs in the SDK if its usable for the general community. It seems like you'd need some kind of rsyslog to Sentry service in addition to this?

Reply:
Yes to the first question, but I was hoping users can directly use the sentry sdk as opposed to an "internal" wrapper.

We don't need a special rsyslog to Sentry service. rsyslog has both a udp and a tcp interface and then supports external sinks such as kafka, so if the appender can talk tcp or udp, there is no other custom requirement. Also the reason, why other vendors such as Loggly have implemented a RSyslogAppender that basically is a UDP client to the rsyslog daemon.

Reply:
Right, but the purpose of this SDK is to get events into the Sentry application. I think any other sink for the Sentry data probably belong in a 3rd party package.